proposals/001-ClusterRepresentation.md

```diff
@@ -0,0 +1,344 @@
+* Ability to add/remove/change virtual or physical clusters at runtime, without interrupting connections established.
+
+
+## Non-Goals
+
+* There's a possiblity of *one-to-many* where a single virtual cluster maps to several physical, presenting them as if they were one. However, supporting transactions across two or more physical clusters would be difficult.  This use-case is out-of-scope.
+* It might be possible for a virtual cluster to present a subset of a physical cluster's brokers as virtual brokers by clevery exposing only those brokers that host topic partitions belonging to that virtual cluster.  This might be advantagous in the case where the physical cluster comprises a very large number of brokers.  We won't consider this use-case as part of this proposal.
+
+
+## Current situation
+
+Kroxyliciousis currently limited to exposing a single broker of a single cluster.
+
+
+## Proposal
+
+The proposal will change some existing concepts and introduce some new ones.  These are discussed first.  Then, we talk about the changes in high level responsibilities.
+
+### Key Objects
+
+
+![alternative text](http://www.plantuml.com/plantuml/proxy?cache=no&src=https://raw.github.com/kroxylicious/design/001-ClusterRepresentation/proposals/001/UML/uml.puml)
+
+
+#### Endpoints
+
+Currently `KafkaProxy` uses the `ProxyConfig` object to get a single address to bind to.  
+
+* For the non-TLS case, multiple endpoints are required to support >1 brokers. 
+* For TLS case, SNI gives a mechanism to route the traffic to >1 brokers even when a single endpoint is used, however there may be performance or operational reasons to favour the ability to define endpoints.
+
+`ProxyConfig` will be replaced by an list of _Endpoint_ objects.  
+
+```yaml
+- name: endpoint name
+  address: bind address/port
+  type: PLAIN|TLS
+```
+
+#### Virtual Clusters
+
+Conceptually the _virtual cluster_ represents the cluster that the kafka client connects to.  The virtual clusters list specifies
+all the kafka clusters are being presented by Kroxylicious. 
+
+A _virtual cluster_ references exactly one _physical cluster_. 
+
+A _virtual cluster_ may define a chain of filters.  This provides zero or more filters that the RPCs will pass through as they traverse the virtual cluster,
+before they reach the physical cluster.
+
+A _virtual cluster_ enumerates the virtual brokers that comprise the virtual cluster.
+
+##### Virtual Brokers
+
+Each broker has a reference to an endpoint that will provide the way in to the broker.   For the SNI case, there is also an addresses matcher.  It is an error
+for more than one broker to share an endpoint unless the endpoint type is TLS and the SNI matching address is defined.
+
+Each virtual broker references exactly one physical broker which must exist within the referenced physical cluster.  It is an error if there is not a
+one-to-one correspondence between physical and virtual brokers.
```

Comment:
By physical broker here I am referring to the model object, rather than the the actual Kafka Cluster. Am I trying to say that the brokers defined by virtual cluster configuration needs to match the brokers the defined by the physical cluster model object.

For a scaling event, after scaling up the number of brokers on the Apache Kafka Cluster, an operator would update both the virtual cluster definition and the physical cluster definition consumed by Kroxy in lock step.

At the heart of this proposal is that the user of kroxy configures kroxy to know about the target cluster topology (i.e. static configuration enumerating the target cluster's brokers) and that configuration can be updated at runtime.    An alternative (not explored by this proposal) is let kroxy discover the broker topology of the target cluster.   Kroxy configuration would just need a bootstrap url of the target cluster.  If we took this option, we would need to require TLS connections at the downstream (client side).  This is so that SNI could be used to distinguish the broker traffic arriving at a single endpoint..